launchpad/src/launchpad.rs

```diff
@@ -1,9 +1,519 @@
+        let redeemed_tickets = self.nr_redeemed_tickets().get();
+        let remaining_tickets = winning_tickets - redeemed_tickets;
+
+        let new_first_winning_ticket_position = prev_first_winning_ticket_position + 1;
+        let new_last_winning_ticket_position =
+            new_first_winning_ticket_position + remaining_tickets - 1;
+
+        let next_generation = self.current_generation().get() + 1;
+
+        let ongoing_operation = self.current_ongoing_operation().get();
+        let mut current_ticket_position = match ongoing_operation {
+            OngoingOperationType::None => new_first_winning_ticket_position,
+            OngoingOperationType::RestartConfirmationPeriod { ticket_position } => ticket_position,
+            _ => return sc_error!("Another ongoing operation is in progress"),
+        };
+
+        let gas_before = self.blockchain().get_gas_left();
+
+        let mut winning_ticket_id = self.shuffled_tickets().get(current_ticket_position);
+        self.set_ticket_status(
+            winning_ticket_id,
+            TicketStatus::Winning {
+                generation: next_generation,
+            },
+        );
+        current_ticket_position += 1;
+
+        let gas_after = self.blockchain().get_gas_left();
+        let gas_per_iteration = gas_before - gas_after;
+
+        while current_ticket_position <= new_last_winning_ticket_position {
+            if self.can_continue_operation(gas_per_iteration) {
+                winning_ticket_id = self.shuffled_tickets().get(current_ticket_position);
+                self.set_ticket_status(
+                    winning_ticket_id,
+                    TicketStatus::Winning {
+                        generation: next_generation,
+                    },
+                );
+                current_ticket_position += 1;
+            } else {
+                self.save_progress(&OngoingOperationType::RestartConfirmationPeriod {
+                    ticket_position: current_ticket_position,
+                })
+            }
+        }
+
+        self.winning_tickets_range().set(&(
+            new_first_winning_ticket_position,
+            new_last_winning_ticket_position,
+        ));
+        self.nr_winning_tickets().set(&remaining_tickets);
+        self.nr_redeemed_tickets().clear();
+
+        self.start_confirmation_period(
+            new_first_winning_ticket_position,
+            new_last_winning_ticket_position,
+        );
+
+        Ok(())
+    }
+
+    fn try_get_ticket_range(&self, address: &Address) -> SCResult<(usize, usize)> {
+        require!(
+            !self.ticket_range_for_address(address).is_empty(),
+            "You have no tickets"
+        );
+
+        Ok(self.ticket_range_for_address(address).get())
+    }
+
+    // storage
+
+    // ticket ID -> address mapping
+    #[storage_mapper("ticketOwners")]
+    fn ticket_owners(&self) -> VecMapper<Self::Storage, Address>;
```

Comment:
You're right, it's not useful (anymore). Initially I didn't have that `ticket_range_for_address`. Now since I always know who the caller is, and users don't need to know their ticket IDs themselves, this is going to be removed.

Good catch.